File: node/src/benchmarks/benchmark-worker.ts
/**
 * Multi-Threaded Benchmark Worker Engine for Node.js Cloud Spanner Benchmarks.
 *
 * Architecture & Execution Model:
 * --------------------------------
 * Node.js runs a single-threaded JavaScript execution model (libuv event loop) per process.
 * To achieve true multi-core CPU scaling comparable to multi-threaded runtimes (Go, Java, Rust)
 * on multi-vCPU machines (such as 4-vCPU GCE VMs), this module runs inside a dedicated
 * `worker_threads` Worker.
 *
 * Key Concepts:
 * 1. Isolated V8 Engine & Event Loop:
 *    Each Worker thread runs in its own V8 isolate with an independent event loop and libuv thread pool.
 * 2. Dedicated Spanner Client Instance:
 *    Each worker initializes its own `Spanner` client instance and gRPC channel pool.
 *    This avoids cross-thread synchronization overhead, lock contention, and event-loop lag.
 * 3. Coordinator-Worker Communication Protocol:
 *    - Main Thread (Coordinator): Responsible for benchmark timing, rate-limiting (TPS), burst schedules,
 *      concurrency throttling, and metric aggregation/OpenTelemetry export.
 *    - Worker Thread: Receives execution commands via MessagePort, executes database operations,
 *      computes high-resolution execution latency using `process.hrtime.bigint()`, and reports results.
 *
 * Message Protocol:
 * -----------------
 * Main -> Worker:
 *   - `{ type: 'execute', taskId: number }`
 *     Request to execute a single benchmark operation.
 *   - `{ type: 'stop' }`
 *     Request to gracefully stop after completing in-flight queries.
 *
 * Worker -> Main:
 *   - `{ type: 'ready', workerId: number }`
 *     Notifies coordinator that client initialization is complete and worker is ready for work.
 *   - `{ type: 'completed', taskId: number, latencyUs: number, success: boolean, error?: string }`
 *     Reports result and high-resolution latency for an executed task.
 */

import {parentPort, workerData} from 'worker_threads';
import {Database} from '@google-cloud/spanner';
import {createSpannerClient} from '../spanner/client';
import {executePointSelect} from './point-select';
import {executeSelectAndUpdate} from './select-update';
import {executeReadLargeResultSet} from './read-large-result-set';
import {executeReadNarrowResultSet} from './read-narrow-result-set';

/**
 * Worker initialization data transferred from the main thread via `workerData`.
 * Note: workerData uses the HTML structured clone algorithm, so values must be serializable
 * primitives or plain objects (no functions, sockets, or complex classes).
 */
export interface BenchmarkWorkerData {
  workerId: number;
  benchmarkType: string;
  projectId: string;
  instanceId: string;
  databaseId: string;
  host?: string;
  tableName: string;
  minId: number;
  maxId: number;
  numRows?: number;
  databaseOptions?: Record<string, any>;
}

if (!parentPort) {
  throw new Error('benchmark-worker must be executed as a worker thread');
}

const data = workerData as BenchmarkWorkerData;

// Each worker thread establishes its own Spanner client
const spanner = createSpannerClient(data.projectId, data.host);
const instance = spanner.instance(data.instanceId);
const database = instance.database(data.databaseId, data.databaseOptions);

type ExecutorFn = (
  database: Database,
  tableName: string,
  minId: number,
  maxId: number,
) => Promise<number | void>;

let executeFn: ExecutorFn;

switch (data.benchmarkType) {
  case 'point-select':
    executeFn = executePointSelect;
    break;
  case 'select-update':
    executeFn = executeSelectAndUpdate;
    break;
  case 'read-large-result-set':
    executeFn = (db: Database) =>
      executeReadLargeResultSet(db, data.numRows || 100000);
    break;
  case 'read-narrow-result-set':
    executeFn = (db: Database) =>
      executeReadNarrowResultSet(db, data.numRows || 200000);
    break;
  default:
    throw new Error(
      `Unsupported benchmark type in worker: ${data.benchmarkType}`,
    );
}

// Track concurrent in-flight tasks assigned to this worker so we can cleanly
// drain all active queries before closing the database and exiting.
let inFlightTasks = 0;
let isStopping = false;

/**
 * Closes Spanner database and client connections cleanly before terminating.
 */
async function cleanup(): Promise<void> {
  try {
    await database.close();
  } catch (e) {
    // Ignore close errors during process teardown
  }
  try {
    await spanner.close();
  } catch (e) {
    // Ignore close errors during process teardown
  }
  process.exit(0);
}

// Handle incoming control and execution messages from the main coordinator thread
parentPort.on('message', async (msg: any) => {
  if (!msg || typeof msg !== 'object') {
    return;
  }

  if (msg.type === 'execute') {
    inFlightTasks++;
    const startTimeNs = process.hrtime.bigint();
    try {
      const customLatencyUs = await executeFn(
        database,
        data.tableName,
        data.minId,
        data.maxId,
      );
      const endTimeNs = process.hrtime.bigint();
      const latencyUs =
        typeof customLatencyUs === 'number' && customLatencyUs > 0
          ? customLatencyUs
          : Number(endTimeNs - startTimeNs) / 1000;

      parentPort?.postMessage({
        type: 'completed',
        taskId: msg.taskId,
        latencyUs,
        success: true,
      });
    } catch (err: any) {
      const endTimeNs = process.hrtime.bigint();
      const latencyUs = Number(endTimeNs - startTimeNs) / 1000;
      parentPort?.postMessage({
        type: 'completed',
        taskId: msg.taskId,
        latencyUs,
        success: false,
        error: err?.message || String(err),
      });
    } finally {
      inFlightTasks--;
      if (isStopping && inFlightTasks === 0) {
        await cleanup();
      }
    }
  } else if (msg.type === 'stop') {
    isStopping = true;
    if (inFlightTasks === 0) {
      await cleanup();
    }
  }
});

// Signal readiness to coordinator thread
parentPort.postMessage({
  type: 'ready',
  workerId: data.workerId,
});

File: python/src/benchmarks/benchmark_worker.py
import sys
import threading
import time
from concurrent.futures import ThreadPoolExecutor
from dataclasses import dataclass
from multiprocessing.connection import Connection
from typing import Any, Callable, Optional, Tuple

from src.benchmarks.point_select import execute_point_select
from src.benchmarks.read_large_result_set import execute_read_large_result_set
from src.benchmarks.read_narrow_result_set import execute_read_narrow_result_set
from src.benchmarks.select_update import execute_select_and_update
from src.benchmarks.tpcc.transactions import execute_tpcc_transaction
from src.spanner.client import create_spanner_client


@dataclass
class BenchmarkWorkerData:
    """Configuration transferred from coordinator to worker process during initialization."""

    worker_id: int
    benchmark_type: str
    project_id: str
    instance_id: str
    database_id: str
    host: Optional[str] = None
    table_name: str = "test"
    min_id: int = 1
    max_id: int = 1000000
    num_rows: int = 100000
    lazy_decode: bool = False
    scale_factor: int = 1
    items: int = 100000
    extended: bool = False
    concurrency: int = 10


def _safe_call(action: Callable[[], Any]) -> None:
    try:
        action()
    except Exception:
        pass


def worker_process_main(
    worker_id: int,
    pipe: Connection,
    worker_data: BenchmarkWorkerData,
) -> None:
    try:
        spanner_client = create_spanner_client(worker_data.project_id, worker_data.host)
        database = spanner_client.instance(worker_data.instance_id).database(
            worker_data.database_id
        )
    except Exception as err:
        print(
            f"Failed to initialize worker {worker_id}: {err}",
            file=sys.stderr,
        )
        try:
            pipe.send(
                {
                    "type": "error",
                    "worker_id": worker_id,
                    "error": str(err),
                }
            )
        except Exception:
            pass
        return

    # For TPC-C, verify database capacity on worker 0 during initialization
    if worker_id == 0 and worker_data.benchmark_type == "tpcc":
        try:
            with database.snapshot(multi_use=False) as snapshot:
                results = snapshot.execute_sql("SELECT COUNT(*) FROM warehouse")
                warehouse_count = 0
                for row in results:
                    warehouse_count = row[0]
                    break
                if warehouse_count < worker_data.scale_factor:
                    raise RuntimeError(
                        f"Database capacity check failed: Required scale factor {worker_data.scale_factor} warehouses, but database only has {warehouse_count}"
                    )
        except Exception as err:
            print(
                f"Worker {worker_id} capacity check failed: {err}",
                file=sys.stderr,
            )
            try:
                pipe.send(
                    {
                        "type": "error",
                        "worker_id": worker_id,
                        "error": str(err),
                    }
                )
            except Exception:
                pass
            return

    # Resolve execution function based on benchmark type
    if worker_data.benchmark_type == "point-select":

        def execute_fn() -> Tuple[None, Optional[float], Optional[str]]:
            execute_point_select(
                database,
                worker_data.table_name,
                worker_data.min_id,
                worker_data.max_id,
            )
            return None, None, None

    elif worker_data.benchmark_type == "select-update":

        def execute_fn() -> Tuple[None, Optional[float], Optional[str]]:
            execute_select_and_update(
                database,
                worker_data.table_name,
                worker_data.min_id,
                worker_data.max_id,
            )
            return None, None, None

    elif worker_data.benchmark_type == "read-large-result-set":

        def execute_fn() -> Tuple[None, Optional[float], Optional[str]]:
            latency_us = execute_read_large_result_set(
                database, worker_data.num_rows, worker_data.lazy_decode
            )
            return None, latency_us, None

    elif worker_data.benchmark_type == "read-narrow-result-set":

        def execute_fn() -> Tuple[None, Optional[float], Optional[str]]:
            latency_us = execute_read_narrow_result_set(
                database, worker_data.num_rows, worker_data.lazy_decode
            )
            return None, latency_us, None

    elif worker_data.benchmark_type == "tpcc":

        def execute_fn() -> Tuple[None, Optional[float], Optional[str]]:
            tx_type = execute_tpcc_transaction(
                database,
                worker_data.scale_factor,
                worker_data.items,
                worker_data.extended,
            )
            return None, None, tx_type

    else:
        print(
            f"Unsupported benchmark type in worker {worker_id}: {worker_data.benchmark_type}",
            file=sys.stderr,
        )
        pipe.send(
            {
                "type": "error",
                "worker_id": worker_id,
                "error": f"Unsupported benchmark type: {worker_data.benchmark_type}",
            }
        )
        return

    pipe_lock = threading.Lock()

    def safe_send(msg: dict) -> None:
        with pipe_lock:
            try:
                pipe.send(msg)
            except (BrokenPipeError, EOFError, OSError):
                pass

    executor = ThreadPoolExecutor(max_workers=max(1, worker_data.concurrency))

    def run_task(task_id: int) -> None:
        start_time = time.perf_counter()
        try:
            _, custom_latency_us, tx_type = execute_fn()
            end_time = time.perf_counter()
            latency_us = (
                custom_latency_us
                if custom_latency_us is not None
                else (end_time - start_time) * 1000000.0
            )
            safe_send(
                {
                    "type": "completed",
                    "task_id": task_id,
                    "worker_id": worker_id,
                    "latency_us": latency_us,
                    "success": True,
                    "tx_type": tx_type,
                    "error": None,
                }
            )
        except Exception as err:
            end_time = time.perf_counter()
            tx_type = getattr(err, "tx_type", None)
            safe_send(
                {
                    "type": "completed",
                    "task_id": task_id,
                    "worker_id": worker_id,
                    "latency_us": (end_time - start_time) * 1000000.0,
                    "success": False,
                    "tx_type": tx_type,
                    "error": str(err),
                }
            )

    # Notify coordinator that client initialization is complete
    safe_send({"type": "ready", "worker_id": worker_id})

    try:
        while True:
            try:
                msg = pipe.recv()
            except (EOFError, KeyboardInterrupt):
                break

            if not isinstance(msg, dict):
                continue

            msg_type = msg.get("type")
            if msg_type == "execute":
                task_id = msg.get("task_id", 0)
                executor.submit(run_task, task_id)
            elif msg_type == "stop":
                break
    finally:
        # Graceful teardown
        executor.shutdown(wait=True)
        _safe_call(
            lambda: (
                database.channel_pool.close()
                if getattr(database, "channel_pool", None)
                else None
            )
        )
        _safe_call(lambda: database.pool.close())
        _safe_call(lambda: database.spanner_api.transport.close())
        _safe_call(lambda: spanner_client.database_admin_api.transport.close())
        _safe_call(lambda: spanner_client.instance_admin_api.transport.close())
        _safe_call(lambda: spanner_client.close())
